### DISCUSSION TITLE
Increase formatter speed

### DISCUSSION THREAD

**Discussion by ydaveluy** [eclipse-langium/langium#1627] [Ideas]
The current formatter implementation returns an array of all text edits, even if the document is already properly formatted and there are no changes to be made.

This array is then returned to VSCode, which applies all the edits.

By filtering the returned edits, overall performance is improved by 20% to 30%.

For example, on a reference DSL with 240,000 lines, formatting time is reduced from 4.6 seconds to 3.5 seconds.

You can see the relevant code here:
https://github.com/eclipse-langium/langium/blob/b2c40e3a8f06becfa7b18b04cfbef7b7931f164a/packages/langium/src/lsp/formatter.ts#L161

The following code shows how to filter the edits:
```ts
if (edit.newText !== textDocument.getText(edit.range))
    edits.push(edit);
```

**Comment by msujew** [comment]
Hey @ydaveluy,

we used to have code like that, see:
https://github.com/eclipse-langium/langium/pull/1550/files

However, this led to a lot of issues in the formatter in case there are conflicting formattings for the same tokens. In that case, the formatter has shown cyclic behavior - everytime you perform the formatting you get a different result. I haven't found a good solution for this yet, so we disabled the behavior again.

We would probably need a post-processing phase after running the `avoidOverlappingEdits` method that filters changes that wouldn't do anything. @ydaveluy Are you interesting in contributing something for that?

**Reply by ydaveluy** [reply]
Instead of filtering the edits when iterating on the AST it can be performed at the end, e.g after filtering overlapping edits.

```ts
    protected avoidOverlappingEdits(textDocument: TextDocument, textEdits: TextEdit[]): TextEdit[] {
        const edits: TextEdit[] = [];
        for (const edit of textEdits) {
            const last = edits[edits.length - 1];
            if (last) {
                const currentStart = textDocument.offsetAt(edit.range.start);
                const lastEnd = textDocument.offsetAt(last.range.end);
                if (currentStart < lastEnd) {
                    edits.pop();
                }
            }
            edits.push(edit);
        }
        return edits.filter(e => e.newText !== textDocument.getText(e.range));
    }
```

**Reply by msujew** [reply]
Yes, that's exactly what I said in https://github.com/eclipse-langium/langium/discussions/1627#discussioncomment-10327492 :)

**Reply by ydaveluy** [reply]
Your comment was edited :-).

Yes I can provide a PR for that.
